### PR TITLE
Initial draft of upgrade guide for kubeadm clusters.

### PR DESCRIPTION
**Pull request by pipejakob**
This is a work-in-progress guide for upgrading kubeadm clusters from 1.6.x to 1.7.x.

Based on SIG feedback, I'm aiming for an initial MVP to get committed so that others can help iterate on it.

Fixes https://github.com/kubernetes/kubeadm/issues/278

@kubernetes/sig-cluster-lifecycle-pr-reviews

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes.github.io/3999)

### CODE REVIEW

**Comment by lukemarsden** [comment]
Thanks! Needs a link from the kubeadm guide to the upgrade guide.

File: docs/tasks/administer-cluster/kubeadm-upgrade-1-7.md [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+---
+
+* TOC
+{:toc}
+
+This guide is for upgrading kubeadm clusters from version 1.6.x to 1.7.x, or
+other 1.6.x releases. Upgrades are not supported before the 1.6.0 release (when
+kubeadm became Beta).
+
+It is a work-in-progress (not intended for merging yet) and targeted at other
```
Comment (timothysc) [inline docs/tasks/administer-cluster/kubeadm-upgrade-1-7.md]: I put the do-not-merge label on so you can remove this, plus the WIP in title also denotes.

File: docs/tasks/administer-cluster/kubeadm-upgrade-1-7.md [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+1. Back up `/etc/kubernetes`.
+
+2. Upgrade system packages.
+
+   Upgrade kubectl, kubeadm, kubelet, and kubernetes-cni via OS packages
+
+   a. On Debian, this would be:
+
+       sudo apt-get update
+       sudo apt-get install kubelet kubeadm kubelet kubernetes-cni
```
Comment (timothysc) [inline docs/tasks/administer-cluster/kubeadm-upgrade-1-7.md]: This should not be needed if it's an upgrade.

File: docs/tasks/administer-cluster/kubeadm-upgrade-1-7.md [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+   b. On CentOS/Fedora, this would be:
+
+       sudo yum update
+       sudo yum install kubelet kubeadm kubelet kubernetes-cni
```
Comment (timothysc) [inline docs/tasks/administer-cluster/kubeadm-upgrade-1-7.md]: same comment as the last one.

File: docs/tasks/administer-cluster/kubeadm-upgrade-1-7.md [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+3. Perform kubeadm upgrade.
+
+       kubeadm init --skip-preflight-checks --kubernetes-version <DESIRED_VERSION>`
+
+   At release time, this would be:
+
+       kubeadm init --skip-preflight-checks --kubernetes-version v1.7.0
+
+   For pre-release testing, this would be:
```
Comment (timothysc) [inline docs/tasks/administer-cluster/kubeadm-upgrade-1-7.md]: Nix from doc, I don't want this publicly visible honestly.

File: docs/tasks/administer-cluster/kubeadm-upgrade-1-7.md [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+       kubeadm init --skip-preflight-checks --kubernetes-version v1.7.0-beta.0
+
+4. Upgrade CNI provider.
+
+   Your CNI provider might have its own upgrade instructions to follow now.
```
Comment (timothysc) [inline docs/tasks/administer-cluster/kubeadm-upgrade-1-7.md]: Links to the major providers?

File: docs/tasks/administer-cluster/kubeadm-upgrade-1-7.md [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+TODO: Rollback instructions in case anything goes wrong (using the backed up
```
Comment (timothysc) [inline docs/tasks/administer-cluster/kubeadm-upgrade-1-7.md]: cp -r backup /etc/kubernetes
yum downgrade x,y,z

File: docs/tasks/administer-cluster/kubeadm-upgrade-1-7.md (hunk 2 above)
Comment (luxas) [inline docs/tasks/administer-cluster/kubeadm-upgrade-1-7.md]: Yeah, `sudo apt-get update && sudo apt-get upgrade` is sufficent

File: docs/tasks/administer-cluster/kubeadm-upgrade-1-7.md (hunk 5 above)
Comment (luxas) [inline docs/tasks/administer-cluster/kubeadm-upgrade-1-7.md]: There is an addons page, we might be able to link to that one...

File: docs/tasks/administer-cluster/kubeadm-upgrade-1-7.md [hunk 7]
```diff
@@ -0,0 +1,56 @@
+---
+assignees:
+- pipejakob
+title: Upgrading kubeadm clusters to 1.7
```
Comment (luxas) [inline docs/tasks/administer-cluster/kubeadm-upgrade-1-7.md]: from v1.6 to v1.7?

File: docs/tasks/administer-cluster/kubeadm-upgrade-1-7.md [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+- "/docs/admin/kubeadm-upgrade-1-7/"
+- "/docs/admin/kubeadm-upgrade-1-7.html"
+---
+
+* TOC
+{:toc}
+
+This guide is for upgrading kubeadm clusters from version 1.6.x to 1.7.x, or
+other 1.6.x releases. Upgrades are not supported before the 1.6.0 release (when
+kubeadm became Beta).
```
Comment (timothysc) [inline docs/tasks/administer-cluster/kubeadm-upgrade-1-7.md]: @luxas going to need to add the warning re: https://github.com/kubernetes/kubernetes/pull/47081#discussion_r120518310

**Comment by pipejakob** [comment]
I've addressed most of the feedback given, so please give this another look.

I still intend to add explicit rollback instructions, and I'll also host developer builds of debs for the most recent 1.7.0 beta release so others can actually test the instructions and file bugs for any problems found.

**Comment by pipejakob** [comment]
I've built debs for `1.7.0-beta.2` for anyone willing to test these instructions, which are hosted in GCS:

    gs://beacham-public/kubernetes/1.7.0-beta.2/debs

There are subfolders for the different architectures, so to grab all of the debs for, say, amd64, you can run:

    mkdir kubernetes-debs
    gsutil rsync gs://beacham-public/kubernetes/1.7.0-beta.2/debs/amd64 kubernetes-debs

When following the instructions, instead of running `sudo apt-get update && sudo apt-get upgrade` you would instead run `sudo dpkg -i kubernetes-debs/*`.

If anyone wants to test the instructions using rpms, please let me know and I'll try to build those as well.

**Comment by pipejakob** [comment]
I've gone ahead and built rpms for testing as well:

    gs://beacham-public/kubernetes/1.7.0-beta.2/rpms

File: docs/tasks/administer-cluster/kubeadm-upgrade-1-7.md [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+1. Back up `/etc/kubernetes`.
+
+2. Upgrade system packages.
+
+   Upgrade your OS packages for kubectl, kubeadm, kubelet, and kubernetes-cni.
+
+   a. On Debian, this can be accomplished with:
+
+       sudo apt-get update
+       sudo apt-get upgrade
```
Comment (timothysc) [inline docs/tasks/administer-cluster/kubeadm-upgrade-1-7.md]: I recommend just the update steps, it will keep around the legacy files in case there is one.

File: docs/tasks/administer-cluster/kubeadm-upgrade-1-7.md [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+
+   b. On CentOS/Fedora, you would instead run:
+
+       sudo yum update
+       sudo yum upgrade
+
```
Comment (timothysc) [inline docs/tasks/administer-cluster/kubeadm-upgrade-1-7.md]: You're missing a 
`sudo systemctl restart kubelet`

File: docs/tasks/administer-cluster/kubeadm-upgrade-1-7.md (hunk 9 above)
Comment (pipejakob) [inline docs/tasks/administer-cluster/kubeadm-upgrade-1-7.md]: I'm confused about the suggestion here. Are you suggesting to delete "sudo apt-get upgrade" and only leave the "sudo apt-get update" step? On Debian, that would only fetch the listing of new packages without upgrading any packages.

File: docs/tasks/administer-cluster/kubeadm-upgrade-1-7.md (hunk 10 above)
Comment (pipejakob) [inline docs/tasks/administer-cluster/kubeadm-upgrade-1-7.md]: Done!

**Comment by pipejakob** [comment]
CC @lukemarsden

**Comment by pipejakob** [comment]
We discussed this PR in this week's SIG Cluster Lifecycle meeting, since it is a hot topic for 1.7 burndown, and we were hoping it could get merged once a few people had tested the steps. If I can get ACKs from others that they've independently verified the process, will I need anything else other than a normal LGTM? I'm not sure how the "Needs Docs Review" label works.

Also, could someone please remove the `do-not-merge` label for me? I don't have permission. Thanks!

File: docs/tasks/administer-cluster/kubeadm-upgrade-1-7.md [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+4. Perform kubeadm upgrade.
+
+       sudo kubeadm init --skip-preflight-checks --kubernetes-version <DESIRED_VERSION>
+
+   For instance, if you want to upgrade to `1.7.0`, you would run:
+
+       sudo kubeadm init --skip-preflight-checks --kubernetes-version v1.7.0
+
+5. Upgrade CNI provider.
```
Comment (luxas) [inline docs/tasks/administer-cluster/kubeadm-upgrade-1-7.md]: The admin has to delete the kube-proxy pods as well (so they get recreated)...
We aren't using rolling updates for that DS yet

Comment (pipejakob) [inline docs/tasks/administer-cluster/kubeadm-upgrade-1-7.md]: Good call-out. I'll add it to the doc and retest the steps.

**Comment by pipejakob** [comment]
> should we say something about how you may be able to restore things?

That was my original intention, but I was taking @lukemarsden's advice and trying to start minimal with the documentation so that we could keep adding to it once it was checked in. Now that it's been in review for so long, and we're so close to the release, I'll hedge a bit and open a second PR based on top of this commit to add the rollback instructions.

File: docs/tasks/administer-cluster/kubeadm-upgrade-1-7.md (hunk 11 above)
Comment (pipejakob) [inline docs/tasks/administer-cluster/kubeadm-upgrade-1-7.md]: Done. However, the 1.7.0-beta.2 version doesn't seem to have the `kube-proxy` regression fix, so the pod ends up crash looping. I tested with kubeadm built from the HEAD of `release-1.7` and the steps worked fine.

The next 1.7 beta release is due out today, so I'll be able to build new debs/rpms for it and post them for testing. I'll also start building them for the HEAD of `release-1.7` so people can still test these instructions ASAP.

**Comment by pipejakob** [comment]
Repeating this comment so it's not hidden in the collapsed, outdated diff:

I've added the step @luxas suggested for manually deleting `kube-proxy` so that it gets recreated during the `kubeadm init`. However, the 1.7.0-beta.2 version doesn't seem to have the kube-proxy regression fix, so the pod ends up crash looping. I tested with kubeadm built from the HEAD of release-1.7 and the steps worked fine.

The next 1.7 beta release is due out today, so I'll be able to build new debs/rpms for it and post them for testing. I'll also start building them for the HEAD of release-1.7 so people can still test these instructions ASAP.

File: docs/tasks/administer-cluster/kubeadm-upgrade-1-7.md (hunk 9 above)
Comment (timothysc) [inline docs/tasks/administer-cluster/kubeadm-upgrade-1-7.md]: My debian is rusty, but on fedora/*EL you don't need the extra step so I'll comment below.

File: docs/tasks/administer-cluster/kubeadm-upgrade-1-7.md [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+
+   a. On Debian, this can be accomplished with:
+
+       sudo apt-get update
+       sudo apt-get upgrade
+
+   b. On CentOS/Fedora, you would instead run:
+
+       sudo yum update
+       sudo yum upgrade
```
Comment (timothysc) [inline docs/tasks/administer-cluster/kubeadm-upgrade-1-7.md]: You don't need the extra upgrade step here, update works just fine and leaves breadcrumb config files in case you clobber something.

File: docs/tasks/administer-cluster/kubeadm-upgrade-1-7.md (hunk 9 above)
Comment (pipejakob) [inline docs/tasks/administer-cluster/kubeadm-upgrade-1-7.md]: Ah, got it. Will remove.

File: docs/tasks/administer-cluster/kubeadm-upgrade-1-7.md (hunk 12 above)
Comment (pipejakob) [inline docs/tasks/administer-cluster/kubeadm-upgrade-1-7.md]: Done.

**Comment by pipejakob** [comment]
For anyone wanting to test the instructions using Debian on amd64, I've created new .debs here:

    gs://beacham-public/kubernetes/ci-1.7.0-beta.ccaf5152ae9f01/debs/amd64/

**Comment by chenopis** [comment]
Ping me when there's a tech lgtm and this is ready to be merged. I'll make sure it gets into the 1.7 release.

**Comment by pipejakob** [comment]
I've added steps to upgrade the OS packages on each node (and restart kubelet), along with headings to make it clear what steps to perform where.

Any other comments, @luxas or @timothysc?

File: docs/tasks/administer-cluster/kubeadm-upgrade-1-7.md [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+**WARNING**: These instructions will **overwrite** all of the resources managed
+by kubeadm (static pod manifest files, service accounts and RBAC rules in the
+`kube-system` namespace, etc.), so any customizations you may have made to these
+resources after cluster setup will need to be reapplied after the upgrade. The
+upgrade will not disturb other static pod manifest files or objects outside the
+`kube-system` namespace.
+
+### On the master
+
+1. Back up `/etc/kubernetes`.
```
Comment (luxas) [inline docs/tasks/administer-cluster/kubeadm-upgrade-1-7.md]: Why should we have this if we don't tell how to restore?

Comment (pipejakob) [inline docs/tasks/administer-cluster/kubeadm-upgrade-1-7.md]: Good point. Ultimately, I'm hoping we can have backup, upgrade, and rollback instructions before the 1.7 release. I've already moved out the rollback instructions to a separate PR, so it probably makes sense to move the backup instructions there as well instead of leaving them dangling.

**Comment by pipejakob** [comment]
@chenopis Tech LGTM is done by @luxas. Care to take a look?
